torch/_lobpcg.py

```diff
@@ -0,0 +1,719 @@
+                 iparams,  # type: Dict[str, int]
+                 fparams,  # type: Dict[str, float]
+                 bparams,  # type: Dict[str, bool]
+                 method,   # type: str
+                 tracker   # type: Optional[None]
+                 ):
+        # type: (...) -> None
+
+        # constant parameters
+        self.A = A
+        self.B = B
+        self.iK = iK
+        self.iparams = iparams
+        self.fparams = fparams
+        self.bparams = bparams
+        self.method = method
+        self.tracker = tracker
+        m = iparams['m']
+        n = iparams['n']
+
+        # variable parameters
+        self.X = X
+        self.E = torch.zeros((n, ), dtype=X.dtype, device=X.device)
+        self.R = torch.zeros((m, n), dtype=X.dtype, device=X.device)
+        self.S = torch.zeros((m, 3 * n), dtype=X.dtype, device=X.device)
+        self.tvars = {}               # type: Dict[str, Tensor]
+        self.ivars = {'istep': 0}     # type: Dict[str, int]
+        self.fvars = {'_': 0.0}       # type: Dict[str, float]
+        self.bvars = {'_': False}     # type: Dict[str, bool]
+
+    def __str__(self):
+        lines = ['LOPBCG:']
+        lines += ['  iparams={}'.format(self.iparams)]
+        lines += ['  fparams={}'.format(self.fparams)]
+        lines += ['  bparams={}'.format(self.bparams)]
+        lines += ['  ivars={}'.format(self.ivars)]
+        lines += ['  fvars={}'.format(self.fvars)]
+        lines += ['  bvars={}'.format(self.bvars)]
+        lines += ['  tvars={}'.format(self.tvars)]
+        lines += ['  A={}'.format(self.A)]
+        lines += ['  B={}'.format(self.B)]
+        lines += ['  iK={}'.format(self.iK)]
+        lines += ['  X={}'.format(self.X)]
+        lines += ['  E={}'.format(self.E)]
+        r = ''
+        for line in lines:
+            r += line + '\n'
+        return r
+
+    def update(self):
+        """Set and update iteration variables.
+        """
+        if self.ivars['istep'] == 0:
+            X_norm = float(torch.norm(self.X))
+            iX_norm = X_norm ** -1
+            A_norm = float(torch.norm(_utils.matmul(self.A, self.X))) * iX_norm
+            B_norm = float(torch.norm(_utils.matmul(self.B, self.X))) * iX_norm
+            self.fvars['X_norm'] = X_norm
+            self.fvars['A_norm'] = A_norm
+            self.fvars['B_norm'] = B_norm
+            self.ivars['iterations_left'] = self.iparams['niter']
+            self.ivars['converged_count'] = 0
+            self.ivars['converged_end'] = 0
+
+        if self.method == 'ortho':
+            self._update_ortho()
+        else:
+            self._update_basic()
+
+        self.ivars['iterations_left'] = self.ivars['iterations_left'] - 1
+        self.ivars['istep'] = self.ivars['istep'] + 1
+
+    def update_residual(self):
+        """Update residual R from A, B, X, E.
+        """
+        mm = _utils.matmul
+        self.R = mm(self.A, self.X) - torch.matmul(mm(self.B, self.X), torch.diag_embed(self.E))
```

Comment:
I'd rather replace `diag_embed` with a point-wise broadcasted multiply.

Reply:
Also, it is cheaper to compute one single matrix product `(A-BD)X`

Reply:
Also, it is cheaper to compute it like `(A - E^T * B)X`.

Reply:
Fixed.

Reply:
A single matrix product now, right? Not sure whether my message got trough though ;)

Reply:
Computing `A - E^T * B` is not a simple option as `A` and `B` may be sparse or dense matrices, or `B` may not be specified at all. Also, if one of the input matrices will be large, one may not have resources to store temporary large matrices.
So, keeping as it is.

Reply:
But you can check when the matrices are sparse and when they are not, am I correct?

Reply:
It turns out that `(A - E^T * B)X` might be more expensive. Here are some benchmark results:
```
(m, k)= (500, 7)
A1, B1 are m x m sparse matrices
A2, B2 are m x m dense matrices
X1 is m x k dense matrix
mm(A1, X1) + mm(B1, X1) took 59.8 us per call
mm(A1 + B1, X1) took 96.1 us per call
mm(A2, X1) + mm(B2, X1) took 31.4 us per call
mm(A2 + B2, X1) took 52.4 us per call

(m, k)= (5000, 7)
A1, B1 are m x m sparse matrices
A2, B2 are m x m dense matrices
X1 is m x k dense matrix
mm(A1, X1) + mm(B1, X1) took 534.666 us per call
mm(A1 + B1, X1) took 988.822 us per call
mm(A2, X1) + mm(B2, X1) took 7156.747 us per call
mm(A2 + B2, X1) took 22027.367 us per call
```

Reply:
Looks like, indeed, working with that mxm chunk of memory is expensive.